apps/client/app/components/PurchaseOrderForm.tsx

```diff
@@ -0,0 +1,177 @@
+  vendor_name: Yup.string().required('Vendor name is required'),
+  order_date: Yup.string().required('Order date is required'),
+  expected_delivery_date: Yup.string().required('Expected delivery date is required').test(
+    'is-greater',
+    'Expected delivery date must be after the order date',
+    function (value) {
+      const { order_date } = this.parent;
+      return new Date(order_date) < new Date(value);
+    }
+  ),
+  purchase_order_line_items: Yup.array().of(
+    Yup.object().shape({
+      id: Yup.number().required('Purchase Order Line Item is required'),
+      item_id: Yup.number().required('Item is required').positive(),
+      quantity: Yup.number().required('Quantity is required').positive().integer(),
+      unit_cost: Yup.number().required('Unit cost is required').positive(),
+    })
+  ).required(),
+  total_quantity: Yup.number().notRequired(),
+  total_cost: Yup.number().notRequired(),
+});
+
+const PurchaseOrderForm: React.FC<PurchaseOrderFormProps> = ({ purchaseOrder, items, mode }) => {
+  const [serverErrors, setServerErrors] = useState<string[]>([]);
+
+  const router = useRouter();
+
+  // Initialize the form with react-hook-form
+  const { control, handleSubmit, formState: { errors } } = useForm<PurchaseOrder>({
+    defaultValues: {
+      id: purchaseOrder?.id || 0,
+      vendor_name: purchaseOrder?.vendor_name || '',
+      order_date: purchaseOrder?.order_date.substring(0, 10) || '',
+      expected_delivery_date: purchaseOrder?.expected_delivery_date.substring(0, 10) || '',
+      purchase_order_line_items: purchaseOrder?.purchase_order_line_items.map(lineItem => ({
+        id: lineItem.id || 0,
+        item_id: lineItem.item_id,
+        quantity: lineItem.quantity,
+        unit_cost: lineItem.unit_cost,
+      })) || [{ id: 0, item_id: 0, quantity: 0, unit_cost: 0 }]
+    },
+    resolver: yupResolver(validationSchema) as any, // Casting to any to get rid of TypeScript error that is not actually an issue
+  });
+
+  const { fields, append, remove } = useFieldArray({
+    control,
+    name: 'purchase_order_line_items'
+  });
+
+  // On form submission, either call the api that will create a Purchase Order or the api that will update a Purchase Order
+  const onSubmit = async (data: Omit<PurchaseOrder, 'id'>) => {
+    const url = mode === Mode.CREATE
+      ? `${API_BASE_URL}/purchase-orders`
+      : `${API_BASE_URL}/purchase-orders/${purchaseOrder?.id}`;
+    const method = mode === Mode.CREATE ? 'POST' : 'PATCH';
+
+    try {
+      const response = await fetch(url, {
+        method,
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(data),
+      });
+
+      if (!response.ok) {
+        const errorData = await response.json();
+        setServerErrors(errorData.message || ['An error occurred']);
+      } else {
+        router.push('/purchase-orders');
+      }
+    } catch (error) {
+      setServerErrors(['An error occurred']);
+    }
```

Comment:
This looks very similar to the other submit above...any thoughts on DRY'ing this up?

Reply:
You can create a utility function like the following and use it in each of the 3 forms: 

`export const handleFormSubmit = async (
  url: string,
  method: string,
  data: any,
  router: any,
  setServerErrors: (errors: string[]) => void,
  redirectPath: string
): Promise<void> => {
  try {
    const response = await fetch(url, {
      method,
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const errorData = await response.json();
      setServerErrors(errorData.message || ['An error occurred']);
    } else {
      router.push(redirectPath);
    }
  } catch (error) {
    setServerErrors(['An error occurred']);
  }
};`